oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtEncoder.java

```diff
@@ -0,0 +1,232 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.jwt;
+
+import java.time.Instant;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSSigner;
+import com.nimbusds.jose.crypto.factories.DefaultJWSSignerFactory;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKMatcher;
+import com.nimbusds.jose.jwk.JWKSelector;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.proc.JWKSecurityContext;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jose.produce.JWSSignerFactory;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithm;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.util.Assert;
+
+/**
+ * Signs and serialized a JWT using the Nimbus library
+ */
+public class NimbusJwtEncoder implements JwtEncoderAlternative {
+	private static final String ENCODING_ERROR_MESSAGE_TEMPLATE = "An error occurred while attempting to encode the Jwt: %s";
+
+	private JwsAlgorithm defaultAlgorithm = SignatureAlgorithm.RS256;
+	private JWKSource<SecurityContext> jwksSource;
+
+	public void setJwkSource(JWKSource<SecurityContext> jwksSource) {
+		Assert.notNull(jwksSource, "jwkSelector cannot be null");
+		this.jwksSource = jwksSource;
+	}
+
+	@Override
+	public NimbusJwtMutator encoder() {
+		Instant now = Instant.now();
+		return new NimbusJwtMutator(this.jwksSource)
+				.jwsHeaders((jws) -> jws
+						.algorithm(this.defaultAlgorithm)
+						.header(JoseHeaderNames.TYP, "JWT")
+				)
+				.claims((claims) -> claims
+						.issuedAt(now)
+						.expiresAt(now.plusSeconds(3600))
+						.notBefore(now)
+				);
+	}
+
+	public static final class NimbusJwtMutator implements JwtMutator<NimbusJwtMutator> {
+		private final JWSSignerFactory jwsSignerFactory = new DefaultJWSSignerFactory();
+		private final JWKSource<SecurityContext> jwkSource;
+		private final NimbusJwtClaimMutator claims = new NimbusJwtClaimMutator();
+		private final NimbusJwsHeaderMutator jwsHeaders = new NimbusJwsHeaderMutator();
+
+		private JWK jwk;
+		private SecurityContext context;
+
+		private NimbusJwtMutator(JWKSource<SecurityContext> jwkSource) {
+			this.jwkSource = jwkSource;
+		}
+
+		@Override
+		public NimbusJwtMutator jwsHeaders(Consumer<JwsHeaderMutator<?>> headersConsumer) {
+			headersConsumer.accept(this.jwsHeaders);
+			return this;
+		}
+
+		/**
+		 * Use this {@link JWK} to sign the JWT
+		 */
+		public NimbusJwtMutator jwsKey(JWK jwk) {
```

Comment:
This is intended to demonstrate that this pattern simplifies exposing library-specific domain objects that we don't want to replicate in Spring Security, like Nimbus's `JWK`.

I think this method might be unnecessary since with `jwsSecurityContext` (see below) the caller could specify a `JWKSecurityContext`. But, I've left it in here for demonstration.